crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -510,91 +551,147 @@ fn host_cancel_read<U, S: AsContextMut<Data = U>>(mut store: S, rep: u32) -> Res
+    // and we can keep it that way.
+    //
+    // If the read state was any other state, then we must set the new state to open
+    // to indicate that there *is* data to be read
+    let new_state = if let ReadState::Closed(read_err_ctx) = &transmit.read {
+        ReadState::Closed(*read_err_ctx)
     } else {
         ReadState::Open
     };
 
+    // Swap in the new read state
     match mem::replace(&mut transmit.read, new_state) {
+        // If the guest was ready to read, then we cannot close the reader (or writer)
+        // we must deliver the event, and update the state associated with the handle to
+        // represent that a read must be performed
         ReadState::GuestReady {
             ty,
             instance,
             handle,
             caller,
             ..
         } => unsafe {
+            // Ensure the final read of the guest is queued, with appropriate closure indicator
             push_event(
                 store,
                 transmit_id.rep(),
                 match ty {
                     TableIndex::Future(_) => Event::FutureRead,
                     TableIndex::Stream(_) => Event::StreamRead,
                 },
-                CLOSED,
+                CLOSED | err_ctx as usize,
                 caller,
             );
 
             *get_mut_by_index(&mut *instance.as_ptr(), ty, handle)?.1 = StreamFutureState::Read;
         },
 
+        // If the the host was ready to read, and the writer end is being closed (host->host write?)
+        // we can accept but discard the write, and close the reader immediately
         ReadState::HostReady { accept } => {
-            accept(Writer::None)?;
-
-            host_close_reader(store, rep)?;
+            accept(Writer::End(err_ctx))?;
+            host_close_reader(store, transmit_rep)?;
         }
 
+        // If the read state is open, then there are no registered readers of the stream/future
+        //
+        // we can delay delivering a final value
         ReadState::Open => {}
 
-        ReadState::Closed => {
-            log::trace!("host_close_writer delete {}", transmit_id.rep());
+        // If the read state was already closed, then we can remove the transmit state completely
+        // (both writer and reader have been closed)
+        ReadState::Closed(_) => {
+            log::trace!("host_close_writer delete {transmit_rep}");
             store.concurrent_state().table.delete(transmit_id)?;
         }
     }
     Ok(())
 }
 
-fn host_close_reader<U, S: AsContextMut<Data = U>>(mut store: S, rep: u32) -> Result<()> {
+/// Close the reader end of a Future or Stream
+///
+/// # Arguments
+///
+/// * `store` - the store for the component
+/// * `transmit_rep` - A global-component-level representation of the transmit state for the reader that should be closed
+/// * `err_ctx` - An optional error context to pass along as the final value of the reader (`0` if none)
```

Comment:
This looks like a cut-and-paste error since there's no `err_ctx` parameter in the signature.